Allow larger strings

Jafaral opened the pull request:
Insrease Max string length from 2^27 to 2^31 on 32-bit platforms
and from 2^31 to 2^63 on 64-bit platforms.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Jafaral committed:
Allow larger strings

Insrease Max string length from 2^31 to 2^59 on 64-bit platforms.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

gtownsend commented:
It would seem from looking at the code that there could be problems allowing the string length to overlap the F_Var bit (among others).  How was this patch tested?

On 13 Jul 2020, at 11:42 AM, Jafar Al-Gharaibeh <notifications@github.com<mailto:notifications@github.com>> wrote:

Insrease Max string length from 2^27 to 2^31 on 32-bit platforms
and from 2^31 to 2^63 on 64-bit platforms.

Signed-off-by: Jafar Al-Gharaibeh to.jafar@gmail.com<mailto:to.jafar@gmail.com>

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/gtownsend/icon/pull/8

Commit Summary

  *   Allow larger strings

File Changes

  *   M src/h/cpuconf.h<https://github.com/gtownsend/icon/pull/8/files#diff-3c83b77050b7e474c309975a236472a7> (4)

Patch Links:

  *   https://github.com/gtownsend/icon/pull/8.patch
  *   https://github.com/gtownsend/icon/pull/8.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/gtownsend/icon/pull/8>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABVOAQHOOHM3TGB2UWO3NPDR3NIQ3ANCNFSM4OYYEM7A>.

Jafaral commented:
@gtownsend   The problem here is two folds
    1- Currently the maximum string size allowed on 64-bit systems is  2^31 -1  (2GB) which is nothing more than an artificial value since even if you want to protect the high bits you would still have access to 59 bits.

   2- As far as I know only the F_Var bit is used with string type, other flag bits are only used with non string types. This should allow us to use bits 27-31 without stepping on the highest F_Var bit. 

Here is my test program:
```
procedure main(args)                                                                                                           |Error (initialization): User jafar has no home directory

    l := 2^args[1] - 1
    s := repl("1", l)
    write(*s)

end  
```
Without my patch on a 32-bit system:
```
root@u32:/opt/icon/bin# ./str 26
67108863
root@u32:/opt/icon/bin# ./str 27

Run-time error 205
File str.icn; Line 4
invalid value
offending value: 134217727
Traceback:
main(list_1 = ["27"])
repl("1",134217727) from line 4 in str.icn
root@u32:/opt/icon/bin# 
````
With my patch:
```
root@u32:/opt/icon/bin# ./str 27
134217727
root@u32:/opt/icon/bin# ./str 28
268435455
root@u32:/opt/icon/bin# ./str 29
536870911
root@u32:/opt/icon/bin# ./str 30
1073741823
root@u32:/opt/icon/bin# ./str 31
2147483647
root@u32:/opt/icon/bin# ./str 32

Run-time error 101
File str.icn; Line 4
integer expected or out of range
offending value: 4294967295
Traceback:
main(list_1 = ["32"])
repl("1",4294967295) from line 4 in str.icn
```
Notice that the error message ow is also more indicative of what is wrong compared to the old behavior. 

On 64-bit systems:
Old behavior:
```
bash-5.0# ./str 27
134217727
bash-5.0# ./str 28
268435455
bash-5.0# ./str 29
536870911
bash-5.0# ./str 30
1073741823
bash-5.0# ./str 31
2147483647
bash-5.0# ./str 32

Run-time error 205
File str.icn; Line 4
invalid value
offending value: 4294967295
Traceback:
main(list_1 = ["32"])
repl("1",4294967295) from line 4 in str.icn

```
New behavior:
```
bash-5.0# ./str 31
2147483647
bash-5.0# ./str 32
4294967295
bash-5.0# ./str 33
8589934591
bash-5.0# ./str 34
17179869183
bash-5.0# ./str 35
34359738367
bash-5.0# 
```
Up to 32GB long string works as expected. I'm not going to test 2^63 long strings any time soon :-)

gtownsend commented:
I presume there was a reason the values were set as they are now.  This looks like a risky patch to me.  If I’m going to spend more time with it I’d like to see

— a test suitable for the Icon test collection that does lots of things with large strings,
including garbage collection

— a convincing argument beyond AFAIK that you’ve examined the code

— in particular, reasoning about why line 747 of interp.r isn’t vulnerable to a problem

Thanks,
Gregg


On 14 Jul 2020, at 12:09 PM, Jafar Al-Gharaibeh <notifications@github.com<mailto:notifications@github.com>> wrote:

@gtownsend<https://github.com/gtownsend> The problem here is two folds

1- Currently the maximum string size allowed on 64-bit systems is 2^31 -1 (2GB) which is nothing more than an artificial value since even if you want to protect the high bits you would still have access to 59 bits.

2- As far as I know only the F_Var bit is used with string type, other flag bits are only used with non string types. This should allow us to use bits 27-31 without stepping on the highest F_Var bit.

Here is my test program:

procedure main(args)                                                                                                           |Error (initialization): User jafar has no home directory

    l := 2^args[1] - 1
    s := repl("1", l)
    write(*s)

end


Without my patch on a 32-bit system:

root@u32:/opt/icon/bin# ./str 26
67108863
root@u32:/opt/icon/bin# ./str 27

Run-time error 205
File str.icn; Line 4
invalid value
offending value: 134217727
Traceback:
main(list_1 = ["27"])
repl("1",134217727) from line 4 in str.icn
root@u32:/opt/icon/bin#


With my patch:

root@u32:/opt/icon/bin# ./str 27
134217727
root@u32:/opt/icon/bin# ./str 28
268435455
root@u32:/opt/icon/bin# ./str 29
536870911
root@u32:/opt/icon/bin# ./str 30
1073741823
root@u32:/opt/icon/bin# ./str 31
2147483647
root@u32:/opt/icon/bin# ./str 32

Run-time error 101
File str.icn; Line 4
integer expected or out of range
offending value: 4294967295
Traceback:
main(list_1 = ["32"])
repl("1",4294967295) from line 4 in str.icn


Notice that the error message ow is also more indicative of what is wrong compared to the old behavior.

On 64-bit systems:
Old behavior:

bash-5.0# ./str 27
134217727
bash-5.0# ./str 28
268435455
bash-5.0# ./str 29
536870911
bash-5.0# ./str 30
1073741823
bash-5.0# ./str 31
2147483647
bash-5.0# ./str 32

Run-time error 205
File str.icn; Line 4
invalid value
offending value: 4294967295
Traceback:
main(list_1 = ["32"])
repl("1",4294967295) from line 4 in str.icn



New behavior:

bash-5.0# ./str 31
2147483647
bash-5.0# ./str 32
4294967295
bash-5.0# ./str 33
8589934591
bash-5.0# ./str 34
17179869183
bash-5.0# ./str 35
34359738367
bash-5.0#


Up to 32GB long string works as expected. I'm not going to test 2^63 long strings any time soon :-)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/gtownsend/icon/pull/8#issuecomment-658360251>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABVOAQGECJHUFEVJ5N4MOTTR3SUPHANCNFSM4OYYEM7A>.

Jafaral commented:
I will look for more details, but at least the 64-bit size limit should be an easy fix.

Jafaral commented:
Hi @gtownsend , 

   I looked back at Icon's implementation book, it does specifically say that the qualifier bit must be set for all not string descriptors. I.e, a descriptor is either a string with a qualifier bit set to zero, or not a string and the bit is set to 1. For things like a variable or pointer descriptors, according to the implementation book, you expect the qualifier bit to be set in addition to the flag bit to indicate it is a variable or a pointer descriptor. However, I checked the actual implementation and found out when doing checks like Var(), which is what you referenced in line 747 in interp.r, only the variable flag bit is checked without confirming whether the descriptor is actually non qualifier. Of course this is not what the actual design calls for but I think it was done intentionally that way as you may suspect as an optimization favoring one less check over allowing rarely used larger string sizes especially in those early days when there wasn't enough RAM to hold large strings anyway.

I think the best course of action here is to leave 32-bit as is, since it is not widely used anyway and just fix the 64-bit implementation to allow up to 59-bits size instead of the current 31-bit limit. I did write programs in recent years with 10GB long strings, so strings bigger than 2GB are not just hypothetical use cases. 

If you are OK with this I will update the PR to drop the 32-bit change and update 64-bit systems to allow up to 59-bits size.

gtownsend commented:
For the most part I’ve been leaving Icon alone in the interest of stability, but raising the string maximum on 64-bit systems (only) has enough utility to be worth the relatively low risk.

I’d still need a new automated test.

(When the IB was written the implementation used 32- or 16-bit words.  There have been at least two supplements that later updated the IB but I haven’t reviewed them in detail.)

Jafaral commented:
> — a test suitable for the Icon test collection that does lots of things with large strings,
> including garbage collection

What kind of a test do you want for 64-bit  large strings ? Even as it is  now, 64-bit works just fine with 2GB strings.  I know for a fact that some Icon users run on systems with only 1GB of RAM. Any practical test I can come up with will not even come close to stressing 64-bit  string allocation and will not be relevant at all to this specific change to increase 64-bit systems string size limit.

gtownsend commented:
I’ve been thinking more about this patch.  I am worried about unforeseen problems in the Icon runtime system if the string size limit is changed.  I’d like to see more analysis of this issue.  What things did you consider, what code was inspected, etc.

I’d also like a convincing demonstration that 33-bit strings work in all contexts.  If 33-bit strings work then I’ll grant the likelihood that longer ones will also work.  I realize that such a demonstration won’t run (with the same parameters) on a small system.

Analysis and testing is a lot more work than changing defined constants.  It’s still necessary.

I’m sorry to be such a wet blanket.  This *seems* like an easy change, and it has value.  But is it really safe?

Gregg

On 15 Jul 2020, at 9:07 AM, Jafar Al-Gharaibeh <notifications@github.com<mailto:notifications@github.com>> wrote:

— a test suitable for the Icon test collection that does lots of things with large strings,
including garbage collection

What kind of a test do you want for 64-bit large strings ? Even as it is now, 64-bit works just fine with 2GB strings. I know for a fact that some Icon users run on systems with only 1GB of RAM. Any practical test I can come up with will not even come close to stressing 64-bit string allocation and will not be relevant at all to this specific change to increase 64-bit systems string size limit.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/gtownsend/icon/pull/8#issuecomment-658844125>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABVOAQCZSHQUPARXC3XICOLR3XH5HANCNFSM4OYYEM7A>.

Jafaral commented:
Hi Gregg,

I agree that the stability of the code is more important than an easy change of a constant. I have done simple examples of allocating large strings > 2GB, but I haven't done enough testing to cover all cases. Given how rare a string of size >2GB would be in practical real world scenarios it probably safer to leave the constants as-is.  I have updated the PR  to drop the 32-bit change and changed the limit to  59 bits on 64-bit systems. I don't have the bandwidth to write or run more tests in the short term given  the low priority nature of this change. Feel free to close this PR for now, we can come back to it if a need arise.

Thanks,
Jafar

gtownsend commented:
I took a look at the code and ran some tests.  I can’t find anything that requires limiting the string length to 32 bits (on a 64-bit system).  In particular I couldn’t find anywhere where it is stored in an int instead of a long.

Furthermore, it turns out that MaxStrLen *only* affects the repl() function and nothing else.

So: I have renamed MaxStrLen to MaxRepl for clarity and raised its value to allow longer repl() results on 64-bit systems.
